DataParallelTable.lua

```diff
@@ -0,0 +1,506 @@
+local gpu_local_copy_buffers = {}
+local base_gpu_index = 1  -- A constant
+
+-- *****************************************************************************
+-- Helper Functions
+-- *****************************************************************************
+-- queryGPUDeviceId - Function to query a tensor or table for the 
+-- GPUID.  For tables we will search the table for CudaTensors, query their 
+-- device and make sure the deviceIds of ALL CudaTensors are on the same GPU.
+local function queryGPUDeviceId(object)
+  if torch.type(object) == 'torch.CudaTensor' then
+    return object:getDevice()
+  end
+
+  local deviceId
+
+  -- Try finding a parameter
+  local stack = {}  -- explicit stack to recurse on tables
+  for key, param in pairs(object) do
+    if key ~= 'modules' then
+      stack[#stack+1] = param  -- Push onto the stack
+    end
+  end
+  while #stack > 0 do
+    local param = stack[#stack]; stack[#stack] = nil  -- Pop the stack
+    if (torch.type(param) == 'table') then
+      for i = 1, #param do stack[#stack+1] = param[i] end  -- Push onto stack
+    elseif (torch.type(param) == 'torch.CudaTensor') then
+      if (torch.numel(param) > 0) then
+        -- Empty tensors are always on GPU "0"
+        local cur_id = param:getDevice()
+        if deviceId == nil then
+          deviceId = cur_id
+        else
+          assert(deviceId == cur_id, 
+            'Found CudaTensor instances from different devices')
+        end
+      end
+    end
+  end
+
+  return deviceId
+end
+
+-- Get an avaliable GPU buffer for asyncGPUCopy.  It is used when the GPU tensor
+-- is not contiguous.
+local function getBuffer()
+  local device = cutorch.getDevice()
+  if not gpu_local_copy_buffers[device] then
+    gpu_local_copy_buffers[device] = torch.CudaTensor()
+  end
+  return gpu_local_copy_buffers[device]
+end
+
+-- setDeviceSafe - Avoid redundant calls to setDevice
+local function setDevice(gpuid)
+  if (cutorch.getDevice() ~= gpuid) then
+    cutorch.setDevice(gpuid)
+  end
+end
+
+-- Asynchronous copy from source to dest from GPU to GPU.
+-- This is borrowed (with modifications) from fbcunn.
+local function asyncGPUCopy(dest, source)
```

Comment:
The new copy has landed. Jono can you update this PR and I'll merge it in? Sorry for the delay